### PR TITLE
Feeders/refactorization for the CLI

### PR DESCRIPTION
**Pull request by albertinisg**
This PR hangs from #116 (which was already LGTM'd)

This PR includes the following changes:
- Split geocoder from restaurant feeder and add geocoding objects in a JSON file (as I mentioned in https://github.com/Fiware/tutorials.TourGuide-App/issues/63#issuecomment-205840078)
- Adapted feeders to work with the CLI interface including new parameters.
- Removed the pre-loaded image. Now all the data should be loaded using the feeders.
- Updated README.md with the feeders documentation.

### CODE REVIEW

File: server/feeders/reservationsgenerator.js [hunk 1]
```diff
@@ -4,111 +4,231 @@
...
+    }
+  });
 
-  console.log('Feeding reservations info in orion.');
-  console.log('Number of restaurants: ' + restaurantsData.length);
+  if (argv.help) {
+    showHelp();
+  }
+
+  if (typeof argv.fromDate === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/reservationsgenerator.js]: Fits in one line.

File: server/feeders/reservationsgenerator.js [hunk 2]
```diff
@@ -4,111 +4,231 @@
...
+      argv.fromDate !== '') {
+    fromDate = argv.fromDate;
+  }
+
+  if (typeof argv.toDate === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/reservationsgenerator.js]: Fits in one line.

File: server/feeders/reservationsgenerator.js [hunk 3]
```diff
@@ -4,111 +4,231 @@
...
+      argv.toDate !== '') {
+    toDate = argv.toDate;
+  }
+
+  if (typeof argv.restaurant === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/reservationsgenerator.js]: Fits in one line.

File: server/feeders/restaurantGeocoder.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      'jsonfile': '../data/restaurants.json',
+      'output': '../data/restaurantsGeo.json'
+    }
+  });
+
+  if (argv.help) {
+    showHelp();
+  }
+
+  if (typeof argv.jsonfile === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/restaurantGeocoder.js]: Fits in one line.

File: server/feeders/restaurantGeocoder.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      argv.jsonfile !== '') {
+    cacheFile = argv.jsonfile;
+  }
+
+  if (typeof argv.output === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/restaurantGeocoder.js]: Fits in one line.

File: server/feeders/restaurantGeocoder.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+/**
+ * Function that handles the Geolocation process of every restaurant
+ *
+ * @param {List} restaurants - List of the restaurants
+*/
+function exportGeocode(restaurants) {
+
+  //restaurantsData = restaurantsData.slice(0,5);
```
Comment (FiwareULPGC) [inline server/feeders/restaurantGeocoder.js]: Commented code.

File: server/feeders/restaurantfeeder.js [hunk 7]
```diff
@@ -4,106 +4,180 @@
...
-          .catch(function(err) {
-            console.log(err);
-          });
-        });
-       }, geoWaitTimeMs);
+  if (argv.help) {
+    showHelp();
+  }
+
+  if (typeof argv.jsonfile === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Fits in one line.

File: server/feeders/restaurantfeeder.js [hunk 8]
```diff
@@ -4,106 +4,180 @@
...
+      argv.jsonfile !== '') {
+    cacheFile = argv.jsonfile;
+  }
+
+  if (typeof argv.geofile === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Fits in one line.

File: server/feeders/restaurantfeeder.js [hunk 9]
```diff
@@ -4,106 +4,180 @@
...
+      argv.geofile !== '') {
+    geolocationFile = argv.geofile;
+  }
+
+  if (typeof argv.restaurant === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Fits in one line.

File: server/feeders/restaurantfeeder.js [hunk 10]
```diff
@@ -4,106 +4,180 @@
...
+    restaurantsAdded++;
+    console.info(restaurantsAdded + '/' + restaurantsData.length);
+  }
+}
+
+/**
+ * Function to read the restaurant information from a given file
+*/
+function loadRestaurantData() {
+
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Unneeded line.

File: server/feeders/restaurantfeeder.js [hunk 11]
```diff
@@ -4,106 +4,180 @@
...
+function loadGeoData() {
+  var data = fs.readFileSync(geolocationFile);
+  geoData = JSON.parse(data);
+}
+
+/**
+ * Function that adapts the information for Orion Context Broker
+*/
+function feedOrionRestaurants() {
+
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Unneeded line.

File: server/feeders/restaurantfeeder.js [hunk 12]
```diff
@@ -121,21 +195,21 @@ var feedOrionRestaurants = function() {
     'municipalityCode': 'postalCode'
   };
 
-  Object.keys(restaurantsData).forEach(function(element, index) {
+  var organization = ['Franchise1', 'Franchise2',
+  'Franchise3', 'Franchise4'
+  ];
+  var capacity = [50, 80, 100, 120, 160, 200];
+
+  restaurantsData.forEach(function(element, index) {
```
Comment (FiwareULPGC) [inline server/feeders/restaurantfeeder.js]: Unneeded line.

File: server/feeders/reviewsgenerator.js [hunk 13]
```diff
@@ -36,133 +38,259 @@ var fiwareHeaders = {
...
-        'value': 'Bitergia'
-      }
-    };
-
-    q.push({
-      'attributes': attr
-    }, returnPost);
-  });
-};
+  if (typeof argv.restaurant === 'string' &&
```
Comment (FiwareULPGC) [inline server/feeders/reviewsgenerator.js]: Fits in one line.

File: server/feeders/reviewsgenerator.js [hunk 14]
```diff
@@ -36,133 +38,259 @@ var fiwareHeaders = {
...
-    })
+    utils.getListByType(RESTAURANT_TYPE, null, fiwareHeaders, queryString)
+    .then(processRestaurants)
     .catch(function(err) {
-      console.log(err.error);
+      console.error(err.error);
     });
-  });
-};
+  }  else if (restaurantSelected) {
```
Comment (FiwareULPGC) [inline server/feeders/reviewsgenerator.js]: May be better to put the if inside {} ? It is hard to follow the lines 146-159.

File: server/idas/ul20.js [hunk 15]
```diff
@@ -286,6 +327,15 @@ function initializeSensor(restaurant, room, type) {
...
+ * Send a new measurement for a sensor.
+ *
+ * @param {String} deviceId - Id of the sensor in the IoT Agent.
+ * @param {String} type - The sensor type.
+ * @param {String} value - The measurement value.
+ * @param {String} servicePath - The servicePath of the sensor.
+ * @return {Promise} Response of the IoT Agent.
+*/
 function updateSensor(deviceId, type, value, servicePath) {
   switch (type) {
```
Comment (FiwareULPGC) [inline server/idas/ul20.js:340]: Indentation.

File: server/idas/ul20.js [hunk 16]
```diff
@@ -235,43 +253,66 @@ function sendObservation(deviceId, data, servicePath) {
...
+ * Initialize a sensor with a default value.
+ *
+ * @param {Object} restaurant - Restaurant entity.
+ * @param {String} room - The room where the sensor will be locaed.
+ * @param {String} type - The type of the new sensor.
+ * @return {Promise} Response of the IoT Agent.
+*/
 function initializeSensor(restaurant, room, type) {
   var deviceId = getDeviceId(restaurant, room, type);
   switch (type) {
```
Comment (FiwareULPGC) [inline server/idas/ul20.js:318]: Indentation.

File: server/utils.js [hunk 17]
```diff
@@ -195,60 +131,82 @@ function fixedEncodeURIComponent(str) {
...
 }
 
+/**
+ * Adds the measurements units to the sensor's schema
+ *
+ * @param {Object} sensor - Sensor to add the units to
+ * @return {Object} sensor - Sensor with the units added
+*/
 function setSchemaUnits(sensor) {
   switch (sensor.additionalType) {
```
Comment (FiwareULPGC) [inline server/utils.js:211]: Indentation.

**Comment by FiwareULPGC** [comment]
We tried to test this PR combining it with the PR with the CLI feature, but we couldn't because the load command is not implemented yet.
load_not_implemented

File: server/feeders/reviewsgenerator.js (hunk 14 above)
Comment (albertinisg) [reply server/feeders/reviewsgenerator.js]: It's already inside brackets :)

**Comment by albertinisg** [comment]
Changes addressed in 6eaa64e2ba401620ad1ee90d7164da4b7ce1a713

As the CLI feature comes in another PR, I suggest to test that there. To test the new feeders, you can follow the new [README.md](https://github.com/albertinisg/tutorials.TourGuide-App/blob/8e7cc37434fd6e6fd637022c8e13b65bd1e8726e/server/feeders/README.md) to load the data.

File: server/feeders/reviewsgenerator.js (hunk 14 above)
Comment (Lenijas) [inline server/feeders/reviewsgenerator.js]: We referred that could be clearer to put all the if block inside curly braces (`{` `}`) instead of use an *elif* like style.

You have:

```
if (organization) {
  ...
}
else if (restaurantSelected) {
  ...
} else {
  ...
}
```

and we suggest:

```
if (organization) {
  ...
}
else {
  if (restaurantSelected) {
    ...
  } else {
    ...
  }
}
```

This is only a code style suggestion, not an error. :)

Comment (FiwareULPGC) [reply server/feeders/reviewsgenerator.js]: We referred that could be clearer to put all the if block inside curly braces (`{` `}`) instead of use an _elif_ like style.

You have:

```
if (organization) {
  ...
}
else if (restaurantSelected) {
  ...
} else {
  ...
}
```

and we suggest:

```
if (organization) {
  ...
}
else {
  if (restaurantSelected) {
    ...
  } else {
    ...
  }
}
```

This is only a code style suggestion, not an error. :)

Comment (albertinisg) [reply server/feeders/reviewsgenerator.js]: All right! Didn't get it before, thanks for the clarification. I've updated the latest commit adding the suggested change :)

**Comment by FiwareULPGC** [comment]
Apart of the behavior with the franchises explained in #130, if it looks good to @jmcanterafonseca, this PR LGTM

**Comment by albertinisg** [comment]
New changes added in a7aec37f7c0eaa362f846cfe54aff6f0a0306edf fixing the Franchises behavior #136

**Comment by dmabtrg** [comment]
I've added a function to fix the franchise of the restaurants upon creation.  The franchise is selected based on the restaurant Id.

**Comment by FiwareULPGC** [comment]
LGTM

**Review by dmabtrg — APPROVED** [review]
LGTM
